### PR TITLE
Driver API cleanups

### PR DESCRIPTION
**Pull request by brendanzab**
Working towards #175 

## TODO

- [x] Clean up existing driver API
- [ ] Move intrinsic registration to driver API
- [x] API docs for driver API
- [x] Move REPL command parsing out of the main parser

### CODE REVIEW

File: crates/pikelet-repl/src/lib.rs [hunk 1]
```diff
@@ -149,15 +141,12 @@ pub fn run(opts: Opts) -> Result<(), Error> {
                     interface.add_history_unique(line.clone());
                 }
 
-                let filename = FileName::virtual_("repl");
-                match eval_print(&mut pikelet, &codemap.add_filemap(filename, line)) {
+                let repl_command = unimplemented!("REPL command parser");
```
Comment (brendanzab) [inline crates/pikelet-repl/src/lib.rs]: I'll need to reimplement the REPL parsing somehow before we can merge this PR

File: crates/pikelet-repl/src/lib.rs [hunk 2]
```diff
@@ -176,86 +171,189 @@ pub fn run(opts: Opts) -> Result<(), Error> {
...
+
+    fn from_str(src: &str) -> Result<ReplCommand, Vec<Diagnostic>> {
+        use combine::char::*;
+        use combine::*;
+
+        let anys1 = || many1(any());
+        let spaces = || skip_many(space());
+        let spaces1 = || skip_many1(space());
+        let ident = || {
+            value(())
```
Comment (Marwes) [inline crates/pikelet-repl/src/lib.rs]: `value` doesn't do anything here.

File: crates/pikelet-repl/src/lib.rs [hunk 3]
```diff
@@ -176,86 +171,189 @@ pub fn run(opts: Opts) -> Result<(), Error> {
...
+
+impl FromStr for ReplCommand {
+    type Err = Vec<Diagnostic>;
+
+    fn from_str(src: &str) -> Result<ReplCommand, Vec<Diagnostic>> {
+        use combine::char::*;
+        use combine::*;
+
+        let anys1 = || many1(any());
+        let spaces = || skip_many(space());
```
Comment (Marwes) [inline crates/pikelet-repl/src/lib.rs]: https://docs.rs/combine/3.6.3/combine/parser/char/fn.spaces.html

File: crates/pikelet-repl/src/lib.rs [hunk 4]
```diff
@@ -176,86 +171,189 @@ pub fn run(opts: Opts) -> Result<(), Error> {
...
+                ),
+                attempt(
+                    choice((attempt(string("type")), string("t")))
+                        .with(spaces1())
+                        .with(anys1())
+                        .map(ReplCommand::TypeOf),
+                ),
+                attempt(
+                    string("let")
+                        .with(spaces())
```
Comment (Marwes) [inline crates/pikelet-repl/src/lib.rs]: spaces1?

File: crates/pikelet-repl/src/lib.rs (hunk 2 above)
Comment (brendanzab) [reply crates/pikelet-repl/src/lib.rs]: Yeah, this was mainly for formatting purposes. Like when [using `succeed identity` in an Elm parser](https://github.com/brendanzab/elm-stlc/blob/770b3bae89315c24d3cc7d80ee4449b0fd94cbab/src/Main.elm#L274-L279).

File: crates/pikelet-repl/src/lib.rs (hunk 4 above)
Comment (brendanzab) [reply crates/pikelet-repl/src/lib.rs]: 😱 you are right!!!

Comment (brendanzab) [reply crates/pikelet-repl/src/lib.rs]: I'm not used to working without a lexer it seems 😅

**Comment by brendanzab** [comment]
I think this is a significant enough improvement to merge, but I'd still like to lift intrinsic registration out of the elaboration context. We should probably make another issue for this.

**Pull request merged**